designs/2023-test-rule-errors/README.md

```diff
@@ -0,0 +1,295 @@
+
+It can be tricky to get schemas to perfectly represent what the allowed input should be, especially when no automated testing exists around this today. Schemas are often too lenient, allowing extra input that the rule implementation ignores or doesn't handle properly. This can result in rules silently misbehaving, not respecting the wishes or intentions of the user's configuration, or just crashing.
+
+By enabling testing of schemas, we'll make it easier to write higher-quality schemas and rules, thus improving the user experience.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+We will augment the error objects of invalid test cases to support a `fatal: true` property. This will indicate that the test case triggered an exception with the given message.
+
+Example with a schema validation error because a non-existent option is passed:
+
+```json
+{
+    "invalid": [
+        {
+            "code": "foo();",
+            "options": [{ "checkFoo": true, "nonExistentOption": true }],
+            "errors": [
+                {
+                    "message": "Value {\"checkFoo\":true,\"nonExistentOption\":true} should NOT have additional properties.",
+                    "fatal": true
+                }
+            ]
+        }
+    ]
+}
+```
+
+To test a custom exception thrown by a rule, simply use the exception text as the message: `"message": "Some custom exception message from the rule."`.
+
+A convenience feature to go along with this is to allow `messageId` to be set to some fixed values like `SCHEMA_VALIDATION_ERROR` or `CUSTOM_RULE_ERROR`. This would be useful since most of the time, the user may not care about the exact text of the exception, but only that the particular type of exception was thrown. This also makes it easier to update the exception text without having to update the messages in all the test cases.
+
+New constraints on test cases with a fatal error:
+
+- `code` is not required (the code may be irrelevant when testing options)
+  - **Open Question**: If this is inconvenient or breaks too many assumptions, we can continue to require it but just let the user provide a dummy value like `code: "foo();"`
+- Only one error object can be provided (can't have multiple exceptions at the same time)
+- Cannot have `output` in the top-level object (no violation/autofix can result from a fatal error)
+- Cannot have `suggestions` in the error object (suggestions are only for violations)
+- This feature can only be used for testing actual user-facing exceptions, not for testing exceptions thrown by the rule tester itself, nor for testing syntax/parsing errors
```

Comment:
> yes please! it'd also be great to have test cases that fail to parse, intentionally

@ljharb thanks for reviewing.

To respond to your comment, I actually didn't intend to allow for testing of syntax/parsing errors, as mentioned on this line. While ESLint will throw an exception on such invalid code, so would any other tool or environment that attempts to run the code, so I don't see why we should be using ESLint to test it. Can you explain why we should include that? I'm open to allowing additional types of exceptions to be tested if we have a good justification for them, but the intention was to limit this feature to these two types:

- ESLint-generated exception during schema validation
- Rule-generated custom exception during rule execution

Note: Explicit validation of valid syntax happens [here](https://github.com/eslint/eslint/blob/3e34418b31664decfb2337de798feafbf985b66c/lib/rule-tester/rule-tester.js#L697) in RuleTester.

Reply:
Because eslint is often ran on code in the process of being typed into an editor. There’s a number of regressions I’ve been unable to test because of this limitation.

Reply:
I don't see how this would be helpful for testing rules, because ESLint doesn't run rules if there's a parsing error.

Reply:
But I don't think rules are even involved when there are parsing errors?

I believe ESLint just [throws an exception](https://github.com/eslint/eslint/blob/3e34418b31664decfb2337de798feafbf985b66c/lib/linter/linter.js#L839) for parsing errors and execution never even reaches the rule.

So rules shouldn't have to concern themselves with invalid syntax. Incomplete code is definitely a concern, but not invalid syntax, to my knowledge.

Reply:
Then I’m not sure how a bunch of people have reported rules crashing on input that doesn’t parse. Perhaps using eslint via api can be made to run on unparseable code?

Reply:
Parsers are expected to return AST or throw an error. If the parser throws an error, ESLint doesn't run rules. There is no AST to work on, so it's impossible to continue.

Those cases you mentioned may be related to parsers that allow syntactically invalid code and return an AST instead of throwing a parsing error.

Reply:
In that case, that's my use case :-) testing my rules against those parsers.

Reply:
Are there any mainstream parsers like that? What's an example of one?

Reply:
I really don't know :-/ but i've had numerous bug reports on eslint-plugin-import, in particular, with partially completed import statements in vscode or other editors, and i've fixed those without a regression test because i couldn't figure out how to test them.

Reply:
One possible existing solution to this problem. Can't you create an instance of `RuleTester` using the parser with special behavior to write test cases against? I've often created multiple sets of test cases for a rule, one with the standard parser, and one with the TypeScript parser, to ensure the rule works in both environments.

Reply:
An example is (currently) `@typescript-eslint/parser`:

https://github.com/typescript-eslint/typescript-eslint/issues/1852

However, I don't see how RuleTester can help with this. If you want a test case for an issue where some parser produces an unexpected AST, then you could simply use that parser in the test case?

Reply:
I suppose i could, if i knew what parser that was :-) but it kind of seems like maybe something vscode-eslint is doing on its own, that's not part of a user's eslint config.

(i already have a utility that i'm in the slow process of extracting to a package that lets a single test case run on up to 5 potential parsers based on eslint, node version, and the language features used in the test case)

Reply:
Okay, so it sounds like the proper solution is to find out what parser the bug occurs under and then test and fix the bug under that parser. Anyway, it doesn't seem like best practice to be writing a fix / test case without being able to reproduce it under the actual parser.

As a result, we shouldn't need to include testing of invalid syntax in this RFC.